---
title: "Dimensionality Reduction - R Methods"
output: html_notebook
---
C. Heiser, 2019

__The purpose of this notebook is to perform dimensionality reduction techniques implemented in the R statistical computing language.__

To benchmark these methods, we are using two single-cell RNA-seq datasets:
* From [Macosko, _et al._ (2015)](https://www.ncbi.nlm.nih.gov/pubmed/26000488), mouse retina scRNA-seq data generated using DropSeq: [GSM1626793](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM1626793)
* From [Herring, _et al._ (2018)](https://www.ncbi.nlm.nih.gov/pubmed/29153838), mouse colon scRNA-seq data generated using inDrops: [GSM2743164](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2743164)

The first set should be __discrete__, containing highly separable cell distributions from the mouse retina, where the latter will be __continuous__, representing hierarchical, branching cell states in the gut epithelium.

Both datasets have been __preprocessed and feature-selected__ to allow for direct input to downstream dimensionality reduction algorithms (rather than first performing PCA to limit dimensional input). These methods are shown in `dev/preprocess_and_filter.ipynb`.

```{r}
setwd('../../scrna2019/') # insert relative path to GLM-PCA directory
source('algs/glmpca.R')
setwd('../furry-couscous/')
source('fcc_utils.r')
library(zinbwave, quietly = T)
library(scRNAseq, quietly = T)
library(SIMLR, quietly = T)
```

Read in feature-selected datasets for testing; both are in cell-by-genes format with cell and gene labels
```{r}
# continuous data - GEO accession # GSM2743164
colon <- read.csv('../inputs/GSM2743164_rep1_colon_rnaseq.processed.tsv', sep='\t', header = T, row.names = 1)
colon.clu <- read.csv('../inputs/colon_clu_seurat.csv', header = F)
# discrete data - GEO accession # GSM1626793
retina <- read.csv('../inputs/GSM1626793_P14Retina_1.processed.tsv', sep='\t', header = T, row.names = 1)
retina.clu <- read.csv('../inputs/retina_clu_seurat.csv', header = F)
```

***
### SIMLR
From [Wang, _et al._ (2017)](https://www.ncbi.nlm.nih.gov/pubmed/28263960), SIMLR is a multikernel learning approach to dimensionality reduction, clustering, and visualization of scRNA-seq counts data. Available on [Bioconductor](https://bioconductor.org/packages/release/bioc/html/SIMLR.html).

First on the colon data.
We can normalize our counts data and feed it into the SIMLR algorithm with the expected number of clusters from Seurat.
```{r}
colon_norm <- arcsinh.norm(colon, margin=1) # normalize by arcsinh-tranforming fractional counts per gene

# perform SIMLR analysis with the estimated number of clusters from prior analysis of data
colon_SIMLR <- SIMLR(t(colon_norm), c = 5, no.dim = 2, normalize = F)
```

```{r}
# plot results
plot.DR(data.frame(colon_SIMLR$ydata), colorby = colon.clu$V1, name='SIMLR')
```

```{r}
# plot results colored by SIMLR's clusters
plot.DR(data.frame(colon_SIMLR$ydata), colorby = colon_SIMLR$y$cluster, name='SIMLR')
```

Perform same workflow on retina data
```{r}
retina_norm <- arcsinh.norm(retina, margin=1) # normalize by arcsinh-tranforming fractional counts per gene

# perform SIMLR analysis with the estimated number of clusters from prior analysis of data
retina_SIMLR <- SIMLR(t(retina_norm), c = 6, no.dim = 2, normalize = F)
```

```{r}
# plot results
plot.DR(data.frame(retina_SIMLR$ydata), colorby = retina.clu$V1, name='SIMLR')
```

```{r}
# plot results colored by SIMLR's clusters
plot.DR(data.frame(retina_SIMLR$ydata), colorby = retina_SIMLR$y$cluster, name='SIMLR')
```

Finally write outputs for 10-D latent space and 2D projections from SIMLR
```{r}
write.csv(colon_SIMLR$F, file = 'Rmethods_out/colon_SIMLR_F.csv', row.names = F)
write.csv(colon_SIMLR$ydata, file = 'Rmethods_out/colon_SIMLR_ydata.csv', row.names = F)

write.csv(retina_SIMLR$F, file = 'Rmethods_out/retina_SIMLR_F.csv', row.names = F)
write.csv(retina_SIMLR$ydata, file = 'Rmethods_out/retina_SIMLR_ydata.csv', row.names = F)
```

***
### ZINB-WaVE
From [Risso, _et al._ (2018)](https://www.nature.com/articles/s41467-017-02554-5), ZINB-WaVE is a zero-inflated negative binomial model for low-dimensional representations of scRNA-seq counts data. Available on [Bioconductor](https://bioconductor.org/packages/release/bioc/html/zinbwave.html).

Create SummarizedExperiment object from counts data, then perform ZINB-WaVE.
```{r}
# transpose matrix to genes x cells format
colon_expt <- SummarizedExperiment(assays = list(counts=data.matrix(t(colon))))

# perform ZINB-WAVE analysis
colon_zinbwave <- zinbwave(colon_expt, K=2, epsilon=1000, verbose=T)
```

Plot resulting 2D embedding:
```{r}
plot.DR(data.frame(colon_zinbwave@reducedDims$zinbwave), colorby = colon.clu$V1, name = 'ZINB-WAVE')
```

Perform same steps on retina data
```{r}
# create SummarizedExperiment object from counts data
# transpose matrix to genes x cells format
retina_expt <- SummarizedExperiment(assays = list(counts=data.matrix(t(retina))))

# perform ZINB-WAVE analysis
retina_zinbwave <- zinbwave(retina_expt, K=2, epsilon=1000, verbose=T)
# plot results
plot.DR(data.frame(retina_zinbwave@reducedDims$zinbwave), colorby = retina.clu$V1, name = 'ZINB-WAVE')
```

Output ZINB-WaVE results to .csv files
```{r}
write.csv(data.frame(colon_zinbwave@reducedDims$zinbwave), file = 'Rmethods_out/colon_ZINB-WAVE.csv', row.names = F) # colon data

write.csv(data.frame(retina_zinbwave@reducedDims$zinbwave), file = 'Rmethods_out/retina_ZINB-WAVE.csv', row.names = F) # retina data
```

***
### GLM-PCA
From [Townes, _et al._ (2019)](https://www.biorxiv.org/content/10.1101/574574v1), GLM-PCA is a generalized linear model-based PCA for dimensionality reduction of scRNA-seq counts data. Available on [GitHub](https://github.com/willtownes/scrna2019).

Let's start with colon dataset again
```{r}
# perform GLM-PCA analysis
colon_glmpca <- glmpca(Y=t(colon), L=2, verbose=T)
# plot results
plot.DR(colon_glmpca$factors, colorby = colon.clu$V1, name='GLM-PCA')
```

Do the same for retina data
```{r}
# perform GLM-PCA analysis
retina_glmpca <- glmpca(Y=t(retina), L=2, verbose=T)
# plot results
plot.DR(retina_glmpca$factors, colorby = retina.clu$V1, name='GLM-PCA')
```

Output GLM-PCA results to .csv files
```{r}
write.csv(colon_glmpca$factors, file = 'Rmethods_out/colon_GLM-PCA.csv', row.names = F) # colon data

write.csv(retina_glmpca$factors, file = 'Rmethods_out/retina_GLM-PCA.csv', row.names = F) # retina data
```
